beers/serializers.py

```diff
@@ -0,0 +1,36 @@
+from rest_framework import serializers
+
+from . import models
+
+class BeerStyleCategorySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = models.BeerStyleCategory
+        fields = '__all__'
+
+
+class BeerStyleTagSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = models.BeerStyleTag
+        fields = '__all__'
+
+
+class BeerStyleSerializer(serializers.ModelSerializer):
+    tags = BeerStyleTagSerializer(many=True)
+    category = BeerStyleCategorySerializer()
+
+    class Meta:
+        model = models.BeerStyle
+        fields = '__all__'
+
+    def create(self, validated_data):
+        tags_data = validated_data.pop('tags')
+        cat_data = validated_data.pop('category')
+        category = models.BeerStyleCategory.objects.create(**cat_data)
```

Comment:
Trouble:

There's no way to create a style without creating a category or tags.

Here's what I think makes sense:

```python

class BeerStyleSerializer(serializers.ModelSerializer):
    tags = BeerStyleTagSerializer(many=True)
    category = BeerStyleCategorySerializer(read_only=True)
    category_id = serializers.PrimaryKeyRelatedField(write_only=True, queryset=models.BeerStyleCategory.objects.all(), required=True, allow_null=False)

    def validate(self, data):
        try:
            data['category'] = validated_data.pop('category_id')
        except KeyError:
            # not specified, which means we're in a PATCH. Don't care
            pass
        return data

    def create(self, validated_data):
        tags = validated_data.pop('tags')
        tag_names = set(i['tag'] for i in tags)
        existing_tags = list(models.BeerStyleTag.objects.filter(tag__in=tag_names)
        missing_tags = tag_names.difference(set(i.tag for i in existing_tags))
        new_tags = list(models.BeerStyleTag.objects.bulk_create(models.BeerStyleTag(tag=i) for i in missing_tags)
        instance = super().create(validated_data)
        instance.tags.set(existing_tags + new_tags)
        return instance

    def update(self, instance, validated_data):
        try:
            tags = validated_data.pop('tags')
        except KeyError:
            return super().update(instance, validated_data)
        tag_names = set(i['tag'] for i in tags)
        existing_tags = list(models.BeerStyleTag.objects.filter(tag__in=tag_names)
        missing_tags = tag_names.difference(set(i.tag for i in existing_tags))
        new_tags = list(models.BeerStyleTag.objects.bulk_create(models.BeerStyleTag(tag=i) for i in missing_tags)
        instance = super().update(instance, validated_data)
        instance.tags.set(existing_tags + new_tags, clear=True)
        return instance
```

That way, the user can specify an existing category without having to create one every time via the `category_id` attribute.

Reply:
Okay, this is what I was having the most trouble with, thanks for the clarification here.  I spent far too long reading the django docs about this.

Reply:
Additional code for validate(self, data):

```python

for low_field, high_field in [('abv_low', 'abv_high'), ('ibu_low', 'ibu_high'), ('og_low', 'og_high'), ('fg_low', 'fg_high')]:
    try:
        low_value = data[low_field]
        high_value = data[high_field]
    except KeyError:
        pass
    if low_value > high_value:
        raise serializers.ValidationError(f'{low_value} cannot be greater than {high_value}')

Reply:
Shameless self-promotion: watch my DRF talk and I go into a bit of detail about this.